### PR TITLE
Add deployment of pipelines

### PR DESCRIPTION
**Pull request by macdiesel**
@edx/pipeline-team PTAL, this is the second pull request, comes after #2

### CODE REVIEW

File: edxpipelines/patterns/__init__.py [hunk 1]
```diff
@@ -0,0 +1,4 @@
+from edxpipelines.patterns import jobs
+from edxpipelines.patterns import piplines
+from edxpipelines.patterns import stages
+from edxpipelines.patterns import tasks
```
Comment (doctoryes) [inline edxpipelines/patterns/__init__.py]:
```
from edxpipelines.patterns import (
    jobs, pipelines, stages, tasks
)
```
?

File: edxpipelines/patterns/__init__.py [hunk 2]
```diff
@@ -0,0 +1,4 @@
+from edxpipelines.patterns import jobs
+from edxpipelines.patterns import piplines
```
Comment (doctoryes) [inline edxpipelines/patterns/__init__.py]: pip => pipe

File: edxpipelines/patterns/piplines.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
+from gomatic import *
```
Comment (doctoryes) [inline edxpipelines/patterns/piplines.py]: piplines => pipelines

File: edxpipelines/pipelines/asg_cleanup.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+import edxpipelines.utils as utils
+
+
+
+@click.command()
+@click.option('--save-config', 'save_config_locally', envvar='SAVE_CONFIG', help='Save the pipeline configuration xml locally', required=False, default=False)
+@click.option('--dry-run', envvar='DRY_RUN', help='do a dry run of  the pipeline installation, and save the pre/post xml configurations locally', required=False, default=False)
+@click.option('--variable_file', 'variable_files', multiple=True, help='path to yaml variable file with a dictionary of key/value pairs to be used as variables in the script', required=False)
+@click.option('-e', '--variable', 'cmd_line_vars', multiple=True, help='key/value of a variable used as a replacement in this script', required=False, type=(str, str), nargs=2)
+def install_pipeline(save_config_locally, dry_run, variable_files=[], cmd_line_vars={}):
```
Comment (doctoryes) [inline edxpipelines/pipelines/asg_cleanup.py]: Better to default these params up above in the `click.option` lines.

File: edxpipelines/pipelines/deploy_ami.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+import edxpipelines.patterns as patterns
+from gomatic import *
+
+
+@click.command()
+@click.option('--save-config', 'save_config_locally', envvar='SAVE_CONFIG', help='Save the pipeline configuration xml locally', required=False, default=False)
+@click.option('--dry-run', envvar='DRY_RUN', help='do a dry run of  the pipeline installation, and save the pre/post xml configurations locally', required=False, default=False)
+@click.option('--variable_file', 'variable_files', multiple=True, help='path to yaml variable file with a dictionary of key/value pairs to be used as variables in the script', required=False)
+@click.option('-e', '--variable', 'cmd_line_vars', multiple=True, help='key/value of a variable used as a replacement in this script', required=False, type=(str, str), nargs=2)
+def install_pipeline(save_config_locally, dry_run, variable_files=[], cmd_line_vars={}):
```
Comment (doctoryes) [inline edxpipelines/pipelines/deploy_ami.py]: As above re: defaults.

File: edxpipelines/pipelines/deploy_ami.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    - asgard_api_endpoints
+    - asgard_token
+    - aws_access_key_id
+    - aws_secret_access_key
+
+    To run this script:
+    python edxpipelines/pipelines/deploy_ami.py --variable_file ../gocd-pipelines/gocd/vars/tools/deploy_edge_ami.yml --variable_file ../gocd-pipelines/gocd/vars/tools/tools.yml
+    python edxpipelines/pipelines/deploy_ami.py --variable_file ../gocd-pipelines/gocd/vars/tools/deploy_edx_ami.yml --variable_file ../gocd-pipelines/gocd/vars/tools/tools.yml
+    python edxpipelines/pipelines/deploy_ami.py --variable_file ../gocd-pipelines/gocd/vars/tools/deploy-mckinsey-ami.yml --variable_file ../gocd-pipelines/gocd/vars/tools/tools.yml
+    """
```
Comment (doctoryes) [inline edxpipelines/pipelines/deploy_ami.py]: Examples FTW!

**Comment by doctoryes** [comment]
First pass complete.

File: edxpipelines/patterns/__init__.py (hunk 1 above)
Comment (macdiesel) [inline edxpipelines/patterns/__init__.py]: I removed this in favor of keeping this file clean and dealing with the imports in a different way.

File: edxpipelines/patterns/__init__.py (hunk 2 above)
Comment (macdiesel) [inline edxpipelines/patterns/__init__.py]: done

File: edxpipelines/patterns/piplines.py (hunk 3 above)
Comment (macdiesel) [inline edxpipelines/patterns/piplines.py]: done

File: edxpipelines/pipelines/asg_cleanup.py (hunk 4 above)
Comment (macdiesel) [inline edxpipelines/pipelines/asg_cleanup.py]: done

File: edxpipelines/pipelines/deploy_ami.py (hunk 5 above)
Comment (macdiesel) [inline edxpipelines/pipelines/deploy_ami.py]: done

File: edxpipelines/patterns/pipelines.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+    Returns:
+        GoCdConfigurator
+    """
+
+    pipeline = configurator \
+        .ensure_pipeline_group(pipeline_group) \
+        .ensure_replacement_of_pipeline(pipeline_name) \
+        .set_git_material(
+            GitMaterial("https://github.com/edx/tubular.git", polling=False, destination_directory="tubular"))
```
Comment (feanil) [inline edxpipelines/patterns/pipelines.py]: Does `polling=False` mean that it won't trigger on changes to the repo or that it won't update the tubular repo at all?

File: README.md [hunk 8]
```diff
@@ -1 +1,30 @@
...
+
+You can then diff these files:
+```
+diff config-before.xml config-after.xml
+```
+
+## Things to look out for
+- Currently any *Secure Variables* must be hashed first by the gocd server before putting them in the script
+- gocd tends to mangle long strings or strings that have carrage returns in them.
+- If you need to view a secure variable in the terminal use ```printf "%s" $ENV_VAR``` as echo will not work.
```
Comment (feanil) [inline README.md]: I think we should remove this line.  If we think a variable is wrong.  We should re-set it to what the correct value is.  If you think there is a need to have this ability, we should put an extra step of setting the secure var to some dummy content before doing this for debugging purposes.

File: edxpipelines/patterns/pipelines.py (hunk 7 above)
Comment (macdiesel) [inline edxpipelines/patterns/pipelines.py]: will not trigger on change.

**Comment by feanil** [comment]
:+1:

**Comment by doctoryes** [comment]
👍

**Pull request merged**